cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java

```diff
@@ -0,0 +1,264 @@
+      }
+      co.cask.cdap.api.lineage.field.TransformOperation merge
+              = new co.cask.cdap.api.lineage.field.TransformOperation(mergeOperationName, mergeDescription,
+                                                                      inputFields, outputs);
+      processedOperations.add(merge);
+    }
+  }
+
+  /**
+   * Create {@link InputField}s from field names which acts as an input to the operation. Creating
+   * InputField requires origin; the name of the operation which created the field. To figure out the
+   * origin, we traverse in the reverse direction from the current stage until we reach source. While
+   * traversing, for each stage we check if the field occurs in any of the output operations recorded
+   * by that stage, if so we have found the origin.
+   *
+   * @param fields the List of input field names for an operation for which InputFields to be created
+   * @param currentStage name of the stage which recorded the operation
+   * @param parents topological ordered list of stages till the current stage
+   * @return List of InputFields
+   */
+  private List<InputField> createInputFields(List<String> fields, String currentStage, List<String> parents) {
+    List<InputField> result = new ArrayList<>();
+    for (String field : fields) {
+      ListIterator<String> branchIterator = parents.listIterator(parents.size());
+      boolean foundOrigin = false;
+      while (branchIterator.hasPrevious() && !foundOrigin) {
+        String stage = branchIterator.previous();
+        // check if the current field is output by any one of operation created by current stage
+        List<Operation> operations = stageOperations.get(stage);
+        ListIterator<Operation> operationsIterator = operations.listIterator(operations.size());
+        while (operationsIterator.hasPrevious() && !foundOrigin) {
+          Operation operation = operationsIterator.previous();
+          switch (operation.getType()) {
+            case READ:
+              ReadOperation read = (ReadOperation) operation;
+              if (read.getOutputFields().contains(field)) {
+                result.add(InputField.of(prefixedOperationName(stage, read.getName()), field));
+                foundOrigin = true;
+              }
+              break;
+            case TRANSFORM:
+              TransformOperation transform = (TransformOperation) operation;
+              if (transform.getOutputFields().contains(field)) {
+                result.add(InputField.of(prefixedOperationName(stage, transform.getName()), field));
+                foundOrigin = true;
+              }
+              break;
+            default:
+              break;
+          }
+        }
+      }
+    }
+    Set<String> stageInputs = stageDag.getNodeInputs(currentStage);
+    if (stageInputs.size() > 1) {
+      return replaceOriginsWithMerge(stageInputs, result);
+    }
+    return result;
+  }
+
+  /**
+   * @return return the topologically sorted nodes till the current stage.
```

Comment:
Is the idea here to get all operations that have a path to this node are its parents? Does topological order guarantee that? I feel like this could produce false positives.

Reply:
true.. fixed the code.